charts/prometheus-blackbox-exporter/templates/ingress.yaml

```diff
@@ -1,35 +1,61 @@
 {{- if .Values.ingress.enabled -}}
-{{- $serviceName := include "prometheus-blackbox-exporter.fullname" . -}}
-{{- $servicePort := .Values.service.port -}}
-{{- $ingressPath := .Values.ingress.path -}}
-{{- if semverCompare ">=1.14.0-0" .Capabilities.KubeVersion.GitVersion }}
+{{- $fullName := include "prometheus-blackbox-exporter.fullname" . -}}
+{{- $svcPort := .Values.service.port -}}
+{{- if and .Values.ingress.className (not (semverCompare ">=1.18-0" .Capabilities.KubeVersion.GitVersion)) }}
+  {{- if not (hasKey .Values.ingress.annotations "kubernetes.io/ingress.class") }}
+  {{- $_ := set .Values.ingress.annotations "kubernetes.io/ingress.class" .Values.ingress.className}}
+  {{- end }}
+{{- end }}
+{{- if semverCompare ">=1.19-0" .Capabilities.KubeVersion.GitVersion -}}
+apiVersion: networking.k8s.io/v1
+{{- else if semverCompare ">=1.14-0" .Capabilities.KubeVersion.GitVersion -}}
 apiVersion: networking.k8s.io/v1beta1
-{{- else }}
+{{- else -}}
 apiVersion: extensions/v1beta1
 {{- end }}
 kind: Ingress
 metadata:
-  name: {{ template "prometheus-blackbox-exporter.fullname" . }}
+  name: {{ $fullName }}
   labels:
-    app.kubernetes.io/name: {{ include "prometheus-blackbox-exporter.name" . }}
-    app.kubernetes.io/instance: {{ .Release.Name }}
-    app.kubernetes.io/managed-by: {{ .Release.Service }}
-    helm.sh/chart: {{ include "prometheus-blackbox-exporter.chart" . }}
+    {{- include "prometheus-blackbox-exporter.labels" . | nindent 4 }}
+  {{- with .Values.ingress.annotations }}
   annotations:
-{{ toYaml .Values.ingress.annotations | indent 4 }}
+    {{- toYaml . | nindent 4 }}
+  {{- end }}
 spec:
+  {{- if and .Values.ingress.className (semverCompare ">=1.18-0" .Capabilities.KubeVersion.GitVersion) }}
```

Comment:
Should we fail if `.Values.ingress.className` but kube < 1.18?

Reply:
Not sure if i get your point. 
ingressClassName is not used if kubeversion >=1.18-0 even if .Values.ingress.className is set.
So thats already what you want, or not?

Reply:
My point is that if you set this value and deploy it in kube < 1.18, it will silently get ignored, it may be better to explicitely warn the user that it is incompatible.

Reply:
I could add something like this to the NOTES.txt.

``` 
{{ if and .Values.ingress.className (semverCompare ">=1.18-0" .Capabilities.KubeVersion.GitVersion) }}
! WARNING !

You've set ".Values.ingressClassName" but it's not supported by your Kubernetes version!
Therefore option was not added in the ingress.yaml.
Please use 'kubernetes.io/ingress.class' annotation instead!
{{ end }}
```

Reply:
@desaintmartin thoughts?

Reply:
Thanks! This way, it is explicit to the user and not silently swallowed.